### ISSUE TITLE
Agree on the `tpm2-rs-base` public module topology

### ISSUE THREAD

**Issue by trent-reed** [tpm-rs/tpm-rs#97]
## Summary

There's a number of ways we could organize this crate, we should agree on what we want the final topology to look like. (Note: This has to do with the publicly visible modules, not the private modules with public re-exports for code organization.)

## Details

As a part of [Discussion #18](https://github.com/orgs/tpm-rs/discussions/18), I've been slowly working on breaking up these monolithic modules. However, there's one last thing we haven't really discussed - and that's the ideal module topology from the perspective of the client.

Keep in mind, the developer module topology (what we see and maintain) does not have to match what the client sees. We can have a completely flat repo from the perspective of the client, but separate into submodules internally and simply re-export.

With that said, I've seen that it's usually nice to expose some kind of a structure to the client, that way there's some sense to the crate layout, and they aren't bombarded with types when they look at the documentation.

We agree that we should break these modules up, but I don't think we agree on what the ideal final state of the modules are from the perspective of the client.

### A Note on Enums in the `constants` Module

I highly suspect that these are only named "constants" in the spec because the folks who penned the spec have the perspective of C programmers.

Enums are constants, according to C. So it makes some kind os sense to say "these types are constants". But from the perspective of rust, they're just `enum` types (well, because of `open_enum`, they're structures with constant `impl` values that look like `enums`, but let's not get too in the weeds over that).

My feeling is these should be treated like other types. So they should be named similarly (`PascalCase`), and if we decide to organize types into a module named `types`, they actually belong there - not in the `constants` module.

If we keep a `constants` module, I think we should literally only have constants in there.

### A Note on Prefix Name Stripping

This has come up a few times ([Issue #65](https://github.com/tpm-rs/tpm-rs/issues/65), [Issue #15](https://github.com/tpm-rs/tpm-rs/issues/15)).

My opinion is that we don't want to separate the prefixes from the type names, though I'm open to discussing this more if folks disagree. I'll copy my reasons for this from [Issue #15](https://github.com/tpm-rs/tpm-rs/issues/15):

> On one hand, it does seem natural to want to namespace things like `tpm2::tpms::AlgProperty`. On the other hand I'd hate to force folks to do `use tpm2::tpms::AlgProperty as TpmsAlgProperty` when they inevetably conflict with `tpm2::tpml::AlgProperty` (or other). This will basically lead to most TPM developers as doing `use tpm2::tpml` and then using `tpml::AlgProperty`, but if that's the case why even have the separate module?
> 
> Also, I'm not necessarily in love with the idea of allowing folks doing `use tpm2::tpms::AlgProperty` when there's basically two different `AlgProperty` types. This feels very easy for folks to want to do, and very much lacks readability.
> 
> The fact that the types are named `<prefix><type>` is an organization strategy that makes sense for the spec, but in natural language what these types actually want to say is something like `tpms::AlgProperty as AlgProperty` (structure) and `tpml::AlgProperty as AlgProperties` (list), but that kind of naming is tough to build standards around - so I definitely don't fault the spec naming choice here.
> 
> 1. I don't want to encourage `use tpm2::tpms::AlgProperty`.
> 2. It feels clear to me that clients will inevitably run into name conflicts since many types share common suffixes, and I don't want to have to force them to `use tpm2::tpms::AlgProperty as TpmsAlgProperty`
> 3. I value alignment with the specification over client-visible module nesting. The TPM is already a complicated piece of technology, aligning as close as possible with the spec on naming for type names is probably a good thing (because I suspect almost everyone will be refering to the spec while doing TPM programming).
> 4. The prefix says something about the type, something interesting that we lose when we chop it off the base typename (it's a "list", it's a "structure", it's a "value", etc). This is something that normally would be expressed in different naming conventions, but we work with what we have.

## Options

With all this out of the way, here are the options as I see them...

### Flat

This structure puts all of the types in the root of `tpm2-rs-base` (though it may structure the types into private submodules for better development organization).

#### Example Client Usage

```rs
use tpm2_rs_base::{
  TPM2_SHA_DIGEST_SIZE, TpmiStCommandTag, TpmsAuthResponse, ...
}
```

#### Pros

*   Pretty simple.

#### Cons

*   All of the types will be in the root, so it creates a big daunting list of types.

### Separation of Constants

This is what we have today.

#### Example Client Usage

```rs
use tpm2_rs_base::{
  constants::{TPM2_SHA_DIGEST_SIZE, ...},
  TpmiStCommandTag, TpmsAuthResponse, ...
}
```

#### Pros

*   Pretty simple.

#### Cons

*   All of the types will be in the root, so it creates a big daunting list of types.
*   I'm not totally sure I see the utility of separating only `constants` into their own submodule.

### Separation of Types from Constants

This is a separation that has nothing in the root, and doesn't add separation per-prefix type.

#### Example Client Usage

```rs
use tpm2_rs_base::{
  constants::{TPM2_SHA_DIGEST_SIZE, ...},
  types::{TpmiStCommandTag, TpmsAuthResponse, ...},
}
```

#### Pros

*   Pretty simple.

#### Cons

*   All of the types will be in `types`, so it creates a big daunting list of types.

### Separation By Prefix

This structure organizes constants into the root of the crate, and separates types into their named submodules.

#### Example Client Usage

```rs
use tpm2_rs_base::{
  TPM2_SHA_DIGEST_SIZE, ...,
  tpmi::{TpmiStCommandTag, ...},
  tpms::{TpmsAuthResponse, ...},
}
```

#### Pros

*   We can add public module-level documentation about the types by their prefix (which may be important to note someplace, and we may not want to repeat in all of the relevant types).

#### Cons

*   May feel a tad redundant since the typename has the same prefix in it.

### Separation By Prefix & Constants

Same as above, except that constants stay in their own module.

#### Example Client Usage

```rs
use tpm2_rs_base::{
  constants::{TPM2_SHA_DIGEST_SIZE, ...},
  tpmi::{TpmiStCommandTag, ...},
  tpms::{TpmsAuthResponse, ...},
}
```

#### Pros

*   We can add public module-level documentation about the types by their prefix (which may be important to note someplace, and we may not want to repeat in all of the relevant types).

#### Cons

*   May feel a tad redundant since the typename has the same prefix in it.
*   I'm not totally sure I see the utility of separating only `constants` into their own submodule.

### Separation By Type & Prefix

This is separation by both `type`/`constants`, and by prefix.

#### Example Client Usage

```rs
use tpm2_rs_base::{
  TPM2_SHA_DIGEST_SIZE, ...
  types::{
    tpmi::{TpmiStCommandTag, ...},
    tpms::{TpmsAuthResponse, ...},
  },
}
```

#### Pros

*   We can add public module-level documentation about the types by their prefix (which may be important to note someplace, and we may not want to repeat in all of the relevant types).

#### Cons

*   May feel a tad redundant since the typename has the same prefix in it.

### Separation By Type & Prefix & Constants

This is separation by both `type`/`constants`, and by prefix.

#### Example Client Usage

```rs
use tpm2_rs_base::{
  constants::{TPM2_SHA_DIGEST_SIZE, ...},
  types::{
    tpmi::{TpmiStCommandTag, ...},
    tpms::{TpmsAuthResponse, ...},
  },
}
```

#### Pros

*   We can add public module-level documentation about the types by their prefix (which may be important to note someplace, and we may not want to repeat in all of the relevant types).

#### Cons

*   May feel a tad redundant since the typename has the same prefix in it.

**Comment by oddcoder** [comment]
> So I'll just omit the solutions that lead to pathing like crate::*::tpm{,i,2b,etc.}::TypeName.

I agree with that since we agreed in coding guidelines to stick with `TypeNames` close to specs as possible. Regarding where put them. I have no preference. I don't mind it currently, bu I will not object to changing either, as long as we do not put every tiny type in a file. I can adapt.
